src/worker.py

```
@@ -2146,23 +2268,145 @@ async def on_fetch(request, env):
-    Silently swallows errors so a notification failure never breaks the
-    parent operation (e.g. grading, peer requests, new assignments).
+    Respects user notification preferences.  Silently swallows errors so a
+    notification failure never breaks the parent operation.
     """
     try:
+        # Check user preferences (default: all enabled)
+        pref_col = _NOTIF_PREF_MAP.get(category, "system_notify")
+        try:
+            pref = await env.DB.prepare(
+                f"SELECT {pref_col} AS enabled FROM notification_preferences"
+                " WHERE user_id = ?"
```

Ruff (0.15.14)

[error] Possible SQL injection vector through string-based query construction

(S608)